pyiceberg/partitioning.py

```diff
@@ -413,8 +413,10 @@ def partition_record_value(partition_field: PartitionField, value: Any, schema:
     the final partition record value.
     """
     iceberg_type = schema.find_field(name_or_id=partition_field.source_id).field_type
-    iceberg_typed_value = _to_partition_representation(iceberg_type, value)
-    transformed_value = partition_field.transform.transform(iceberg_type)(iceberg_typed_value)
+    if not isinstance(value, int):
+        # When adding files, it can be that we still need to convert from logical types to physical types
+        value = _to_partition_representation(iceberg_type, value)
+    transformed_value = partition_field.transform.transform(iceberg_type)(value)
```

Comment:
This is causing bugs, I'm going to revisit this to fix it properly

Reply:
Yes, so I got to the bottom of it. It has to do with the return types of the transforms. eg. When we apply the `bucket` transform, the result is always an int, which is great. The problem is with the identity transform where the destination type is equal to the source type. So when a date comes in, it also comes out.

I think in the end it is better to remove the `_to_partition_representation` and see if we can consolidate this somewhere, but that's a different PR.

Reply:
> So when a date comes in, it also comes out.

is it due to not having support for `datetime` literal? #1542

Reply:
also if its just for adding files, perhaps we can do something special just for that path

Reply:
> also if its just for adding files, perhaps we can do something special just for that path

Yes, that's exactly what I went for. I think we can simplify the logic in subsequent PRs :)